### PR TITLE
Rework WPS332: allow := in comprehensions

### LINKED ISSUE #3121: Rework `WPS332`: allow `:=` in comprehensions
This code should be allowed:

```python
users = [user for user_id in user_ids if (user := get_user(user_id)) is not None]
```

We only allow `:=` in comprehensions / generator expressions now.

### PR DESCRIPTION
**Pull request by milssky**
# I have made things!

## Checklist

- [x] I have double checked that there are no unrelated changes in this pull request (old patches, accidental config files, etc)
- [x] I have created at least one test case for the changes I have made
- [x] I have updated the documentation for the changes I have made
- [x] I have added my changes to the `CHANGELOG.md`

- Closes #3121

### CODE REVIEW

File: tests/test_visitors/test_ast/test_operators/test_walrus.py [hunk 1]
```diff
@@ -18,6 +18,13 @@
     if y > 2
 ]
 """
+correct_walrus_comprehension = """
+some = [
+    x + y
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_operators/test_walrus.py:23]: Please, also test that you can use `:=` in the expression, like `{(key := compute_key(i)): values[key] for i in range(10)}`

File: wemake_python_styleguide/visitors/ast/operators.py [hunk 2]
```diff
@@ -225,6 +225,22 @@ def visit_NamedExpr(
         self,
         node: ast.NamedExpr,
     ) -> None:
-        """Disallows walrus ``:=`` operator."""
-        self.add_violation(consistency.WalrusViolation(node))
+        """Disallows walrus ``:=`` operator outside comprehension."""
+        closest = walk.get_closest_parent(
+            node,
+            (
+                ast.ListComp,
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/ast/operators.py:232]: Please, move this to be a class level constant.

File: wemake_python_styleguide/visitors/ast/operators.py [hunk 3]
```diff
@@ -225,6 +225,22 @@ def visit_NamedExpr(
...
+                ast.SetComp,
+                ast.DictComp,
+                ast.GeneratorExp,
+            )
+        )
+
+        if closest is None:
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/ast/operators.py:239]: Let's use early returns 

```suggestion
        if closest is not None:
            return
```

File: wemake_python_styleguide/visitors/ast/operators.py [hunk 4]
```diff
@@ -225,6 +225,22 @@ def visit_NamedExpr(
         self,
         node: ast.NamedExpr,
     ) -> None:
-        """Disallows walrus ``:=`` operator."""
-        self.add_violation(consistency.WalrusViolation(node))
+        """Disallows walrus ``:=`` operator outside comprehension."""
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/ast/operators.py:228]:
```suggestion
        """Disallows walrus ``:=`` operator outside comprehensions."""
```

File: wemake_python_styleguide/visitors/ast/operators.py (hunk 2 above)
Comment (milssky) [reply wemake_python_styleguide/visitors/ast/operators.py:232]: Do you mean a whole tuple with possible comprehensions?

Comment (sobolevn) [reply wemake_python_styleguide/visitors/ast/operators.py:232]: Yes, like we do with other `ClassVar[AnyNodes]`

Comment (milssky) [reply wemake_python_styleguide/visitors/ast/operators.py]: Pushed :)

**Review by sobolevn — APPROVED** [review]
Great! Thank you!

**Comment by milssky** [comment]
Merged all new changes from master. Can try again? =)

File: wemake_python_styleguide/visitors/ast/operators.py [hunk 5]
```diff
@@ -223,10 +223,29 @@ def _check_string_concat(
...
+    )
+
     def visit_NamedExpr(
         self,
         node: ast.NamedExpr,
     ) -> None:
-        """Disallows walrus ``:=`` operator."""
-        self.add_violation(consistency.WalrusViolation(node))
+        """Disallows walrus ``:=`` operator outside comprehensions."""
+        closest = walk.get_closest_parent(
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/ast/operators.py:238]: Please, create a method for that. Let's call it `_check_walrus_in_comprehension`

**Review by sobolevn — APPROVED** [review]

**Pull request merged**